SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp

```diff
@@ -244,12 +247,78 @@ TArray<UClass*> GetAllSupportedClasses()
 	return Classes.Array();
 }
 
-bool SpatialGDKGenerateSchema()
+void DeleteGeneratedSchemaFiles()
+{
+	const FString SchemaOutputPath = GetDefault<USpatialGDKEditorSettings>()->GetGeneratedSchemaOutputFolder();
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.DirectoryExists(*SchemaOutputPath))
+	{
+		if (!PlatformFile.DeleteDirectoryRecursively(*SchemaOutputPath))
+		{
+			UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Could not clean the generated schema directory '%s'! Please make sure the directory and the files inside are writeable."), *SchemaOutputPath);
+		}
+	}
+	PlatformFile.CreateDirectory(*SchemaOutputPath);
+}
+
+void InitClassPathToSchemaMap()
 {
-	ClassPathToSchema.Empty();
+	TSoftObjectPtr<USchemaDatabase> SchemaDatabasePtr(FSoftObjectPath(TEXT("/Game/Spatial/SchemaDatabase.SchemaDatabase")));
+	SchemaDatabasePtr.LoadSynchronous();
+	const USchemaDatabase* const SchemaDatabase = SchemaDatabasePtr.Get();
 
-	SchemaGeneratedClasses = GetAllSupportedClasses();
+	if (SchemaDatabase)
+	{
+		ClassPathToSchema = SchemaDatabase->ClassPathToSchema;
+		NextAvailableComponentId = SchemaDatabase->NextAvailableComponentId;
+	}
+	else
+	{
+		NextAvailableComponentId = SpatialConstants::STARTING_GENERATED_COMPONENT_ID;
+		// As a safety precaution, if the SchemaDatabase.uasset doesn't exist then make sure the schema generated folder is cleared as well. 
+		DeleteGeneratedSchemaFiles();
+	}
+}
 
+void PreProcessSchemaMap()
+{
+	TArray<FString> EntriesToRemove;
+	for (const auto& EntryIn : ClassPathToSchema)
+	{
+		const FString ClassPath = EntryIn.Key;
+		const FSoftObjectPath ItemToReference(ClassPath);
+		
+		// First check if the object is already loaded into memory.
+		UObject* const ResolvedObject = ItemToReference.ResolveObject();
+		UClass*  const LoadedClass    = ResolvedObject ? nullptr : Cast<UClass>(ItemToReference.TryLoad());
+		
+		const bool bUniqueEntry = AdditionalSchemaGeneratedClasses.Find(LoadedClass) == INDEX_NONE;
```

Comment:
Was this ever not true? Shouldn't ClassPathToSchema only contain unique values?

Reply:
No, it was always returning true because the TArray::Find func return -1 if it doesn't find the item. -1, when converted to a boolean, returns true.

The new statement fixes this issue and also ensures that AdditionalSchemaGeneratedClasses only contains unique values.

Reply:
My question is why would we be adding duplicate values to the array? What's feeding it should contain unique values.